ambry-cloud/src/main/java/com.github.ambry.cloud/azure/CosmosChangeFeed.java

```diff
@@ -0,0 +1,188 @@
+     * @param azureRequestId a random UUID which uniquely identifies each cached info.
+     * @param fetchedEntries {@link List} of cached {@link CloudBlobMetadata} objects.
+     */
+    ChangeFeedCacheEntry(String startContinuationToken, String endContinuationToken, String azureRequestId,
+        List<CloudBlobMetadata> fetchedEntries) {
+      this.startContinuationToken = startContinuationToken;
+      this.endContinuationToken = endContinuationToken;
+      this.azureRequestId = azureRequestId;
+      this.fetchedEntries = fetchedEntries;
+    }
+
+    /**
+     * Return start continuation token.
+     * @return start continuation token.
+     */
+    String getStartContinuationToken() {
+      return startContinuationToken;
+    }
+
+    /**
+     * Return the end continuation token.
+     * @return end continuation token.
+     */
+    String getEndContinuationToken() {
+      return endContinuationToken;
+    }
+
+    /**
+     * Return the azure request id.
+     * @return azure request id.
+     */
+    String getAzureRequestId() {
+      return azureRequestId;
+    }
+
+    /**
+     * Return the fetch entries list.
+     * @return {@link List} of {@link CloudBlobMetadata} entries.
+     */
+    List<CloudBlobMetadata> getFetchedEntries() {
+      return fetchedEntries;
+    }
+  }
+
+  private final ConcurrentHashMap<String, ChangeFeedCacheEntry> changeFeedCache;
+  private final int defaultCacheSize;
+  private final CosmosDataAccessor cosmosDataAccessor;
+
+  /**
+   * Constructor to create a {@link CosmosChangeFeed} object.
+   * @param cacheSize default number of cachedEntries for each partition.
+   * @param cosmosDataAccessor {@link CosmosDataAccessor} object.
+   */
+  public CosmosChangeFeed(int cacheSize, CosmosDataAccessor cosmosDataAccessor) {
+    this.defaultCacheSize = cacheSize;
+    changeFeedCache = new ConcurrentHashMap<>();
+    this.cosmosDataAccessor = cosmosDataAccessor;
+  }
+
+  /**
+   * Get next set of change feed entries for the specified partition, after the {@code azureFindToken}.
+   * The number of entries is capped by maxEntriesSize.
+   * This method creates a cache for change feed entries. If the {@code azureFindToken} is not valid,
+   * or if all the items in the cache are consumed, then it queries cosmos for new entries.
+   * @param azureFindToken {@link AzureFindToken} after which the next entries have to be returned.
+   * @param results {@link List} of {@link CloudBlobMetadata} objects which will be populated by new entries.
+   * @param maxEntriesSize maximum size of all the blobs returned in {@code results}
+   * @param partitionId Partition for which change feed entries have to be returned.
+   * @return updated {@link AzureFindToken} after processing the next set of entries.
+   */
+  public AzureFindToken getNextEntriesAndToken(AzureFindToken azureFindToken, List<CloudBlobMetadata> results,
```

Comment:
I find this class confusing, especially this getter method.  It definitely feels like a cache, so maybe the class name should say so.  (I saw you mentioned it in the javadoc.) I originally thought this class would be used to actually run the change feed queries as opposed to delegating that to the data accessor and holding the results.

Reply:
I have renamed this to CosmosChangeFeedCache.
The reason I have delegated query to CosmosDataAccessor is that I want to keep all the calls to cosmos apis at one place.